pilot/pkg/model/telemetry.go

```diff
@@ -205,68 +210,84 @@ func (t *Telemetries) Tracing(proxy *Proxy) *TracingConfig {
 	ct := t.applicableTelemetries(proxy)
 
 	providerNames := t.meshConfig.GetDefaultProviders().GetTracing()
-	for _, m := range ct.Tracing {
-		currentNames := getProviderNames(m.Providers)
-		// If we set providers are current level, use that. Otherwise, keep parent providers
-		if len(currentNames) > 0 {
-			providerNames = currentNames
-		}
-	}
-	if len(providerNames) == 0 {
-		// No providers set at all, use fallback
+	hasDefaultProvider := len(providerNames) > 0
+
+	if len(ct.Tracing) == 0 && !hasDefaultProvider {
 		return nil
 	}
-	if len(providerNames) > 1 {
-		// User can set multiple, but we don't actually support it, so we will pick the first one
-		log.Debugf("invalid tracing configure; only one provider supported: %v", providerNames)
-	}
-	supportedProvider := providerNames[0]
 
-	cfg := TracingConfig{
-		Provider:                     t.fetchProvider(supportedProvider),
-		UseRequestIDForTraceSampling: true,
-	}
-	if cfg.Provider == nil {
-		cfg.Disabled = true
-		return &cfg
+	clientSpec := TracingSpec{UseRequestIDForTraceSampling: true}
+	serverSpec := TracingSpec{UseRequestIDForTraceSampling: true}
+
+	if hasDefaultProvider {
+		// todo: what do we want to do with more than one default provider?
+		// for now, use only the first provider.
+		fetched := t.fetchProvider(providerNames[0])
+		clientSpec.Provider = fetched
+		serverSpec.Provider = fetched
 	}
+
 	for _, m := range ct.Tracing {
 		names := getProviderNames(m.Providers)
 
-		// We need to figure out if the tracing config at this level is relevant. For example, if we selected
-		// provider X but this config applies to provider Y, we should ignore it.
-		includeConfig := false
-		// If there is no provider set it applies to the default.
-		if len(names) == 0 {
-			includeConfig = true
-		}
-		// Otherwise, make sure one of the configured providers is our selected provider.
-		for _, n := range names {
-			if n == supportedProvider {
-				includeConfig = true
-				break
+		specs := []*TracingSpec{&clientSpec, &serverSpec}
+		if m.Match != nil {
+			switch m.Match.Mode {
+			case tpb.WorkloadMode_CLIENT:
+				specs = []*TracingSpec{&clientSpec}
+			case tpb.WorkloadMode_SERVER:
+				specs = []*TracingSpec{&serverSpec}
 			}
 		}
-		if !includeConfig {
-			break
+
+		if len(names) > 0 {
+			// NOTE: we only support a single provider per mode
```

Comment:
add validation in webhook?

Reply:
I believe we have validation warnings for this scenario currently: https://github.com/istio/istio/blob/master/pkg/config/validation/validation.go#L3604